### PR TITLE
feat: extend8s and extend16s

### PR DESCRIPTION
**Pull request by perseus34x**
ExtendChip handles two Wasm unary ops—i32.extend8_s and i32.extend16_s. Each row has selectors (is_extend8s, is_extend16s), the input word b and result word a as bytes [b0,b1,b2,b3] / [a0,a1,a2,a3], and a single sign bit msb used by the active opcode. Only one selector may be 1; padding rows have is_real=0 and we force msb=0.

 We use one fixed 256-row byte table ByteOpcode::MSB that maps a byte → its most-significant bit. The chip sends exactly one lookup per real row: for extend8_s we query b0, for extend16_s we query b1. This certifies both that the queried operand is a real byte and that msb ∈ {0,1} equals its bit7. Then we enforce the result bytes (BabyBear-safe, no big constants):
    •   extend8_s: a0=b0, a1=a2=a3=0xFF·msb.
    •   extend16_s: a0=b0, a1=b1, a2=a3=0xFF·msb.

 The lookup pins the correct sign bit from the correct byte, and the byte equalities uniquely fix a to the two’s-complement sign-extended value. Selectors gate constraints so only the active path applies; padding rows don’t emit lookups and have msb=0. This avoids redundant bit-decompositions, keeps coefficients in [0,255], closes the cross-table cumulative sum, and handles edge cases like 0x..80 → 0xFFFF_FF80 and 0x..8000 → 0xFFFF_8000 exactly.

## PR Checklist

- [X] Added Tests
- [X] Added Documentation
- [X] Breaking changes

### CODE REVIEW

File: crates/rwasm/machine/src/alu/extend/mod.rs [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+        cols.b = event.b.into();
+        cols.is_extend8s = F::from_bool(event.opcode == Opcode::I32Extend8S);
+        cols.is_extend16s = F::from_bool(event.opcode == Opcode::I32Extend16S);
+
+        let b = event.b;
+        let b_bytes = b.to_le_bytes();
+
+        // Emit one lookup per active opcode and set the unified `msb` column.
+        if event.opcode == Opcode::I32Extend8S {
+            let msb = (b_bytes[0] >> 7) & 1;
```
Comment (djadjka) [inline crates/rwasm/machine/src/alu/extend/mod.rs]: you don't need &1 here, becasue b_bytes already 8 bits

File: crates/rwasm/machine/src/alu/extend/mod.rs [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+        builder.send_byte(
+            ByteOpcode::MSB.as_field::<AB::F>(),
+            local.msb,          // Certified result
+            b[1],               // op1: byte being checked
+            AB::Expr::zero(),   // op2
+            local.is_extend16s, // Multiplicity
+        );
+
+        // i32.extend8_s result bytes
+        builder.when(local.is_extend8s).assert_zero(a[0] - b[0]);
```
Comment (djadjka) [inline crates/rwasm/machine/src/alu/extend/mod.rs]: assert_eq

File: crates/rwasm/machine/src/alu/extend/mod.rs [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+
+        builder.assert_bool(local.is_extend8s);
+        builder.assert_bool(local.is_extend16s);
+        builder.assert_bool(local.msb);
+
+        let is_real = local.is_extend8s + local.is_extend16s;
+        builder.assert_bool(is_real.clone());
+
+        let ff = AB::Expr::from_canonical_u32(0xFF);
+        let a = &local.a.0;
```
Comment (djadjka) [inline crates/rwasm/machine/src/alu/extend/mod.rs]: Word support indexing so you can directly call local.a[0] local.b[1]

File: crates/rwasm/machine/src/alu/extend/mod.rs [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+            ByteOpcode::MSB.as_field::<AB::F>(),
+            local.msb,          // Certified result
+            b[1],               // op1: byte being checked
+            AB::Expr::zero(),   // op2
+            local.is_extend16s, // Multiplicity
+        );
+
+        // i32.extend8_s result bytes
+        builder.when(local.is_extend8s).assert_zero(a[0] - b[0]);
+        builder.when(local.is_extend8s).assert_zero(a[1] - local.msb * ff.clone());
```
Comment (djadjka) [inline crates/rwasm/machine/src/alu/extend/mod.rs]: it's better to use builder notation builder.when(local.is_extend8s).when(local.msb).assert_eq(a[2], ff.clone())

File: crates/rwasm/machine/src/alu/extend/mod.rs [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+        builder.when(local.is_extend8s).assert_zero(a[2] - local.msb * ff.clone());
+        builder.when(local.is_extend8s).assert_zero(a[3] - local.msb * ff.clone());
+
+        // i32.extend16_s result bytes
+        builder.when(local.is_extend16s).assert_zero(a[0] - b[0]);
+        builder.when(local.is_extend16s).assert_zero(a[1] - b[1]);
+        builder.when(local.is_extend16s).assert_zero(a[2] - local.msb * ff.clone());
```
Comment (djadjka) [inline crates/rwasm/machine/src/alu/extend/mod.rs]: for a2 a3 you can use builder.when(local.msb).assert_eq(a[2], ff.clone()) builder.when(local.msb).assert_eq(a[3], ff.clone())

File: crates/rwasm/machine/src/alu/extend/mod.rs [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+        builder.when(local.is_extend16s).assert_zero(a[3] - local.msb * ff.clone());
+
+        // Ensure that `msb` is zero when no extend operation is active.
+        builder.assert_zero(local.msb * (is_real.clone() - AB::Expr::one()));
```
Comment (djadjka) [inline crates/rwasm/machine/src/alu/extend/mod.rs]: I'm not sure that this constraint has any sense, if malefactor fill trace incorrectly, it his problem ) 
Our mission is: if is_real we should build path from b and c to a, the remaining cases do not interest us

**Review by djadjka — CHANGES_REQUESTED** [review]

**Review by djadjka — CHANGES_REQUESTED** [review]
Nice!

File: crates/rwasm/machine/src/alu/extend/mod.rs (hunk 4 above)
Comment (perseus34x) [reply crates/rwasm/machine/src/alu/extend/mod.rs]: builder.when(local.is_extend16s).when(local.msb).assert_eq(a[2], ff); This form is **unsound** because it fails to apply any constraint on a[2] when **msb is 0**. A malicious prover could exploit this gap.

File: crates/rwasm/machine/src/alu/extend/mod.rs (hunk 5 above)
Comment (perseus34x) [reply crates/rwasm/machine/src/alu/extend/mod.rs]: builder.when(local.is_extend16s).when(local.msb).assert_eq(a[2], ff); This form is unsound because it fails to apply any constraint on a[2] when msb is 0. A malicious prover could exploit this gap.

File: crates/rwasm/machine/src/alu/extend/mod.rs (hunk 6 above)
Comment (perseus34x) [reply crates/rwasm/machine/src/alu/extend/mod.rs]: done

**Review by nagatoism — APPROVED** [review]
The constraint are quite straightforward so I guess I have not much to comment.
